tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/tracee/UJsonTraceReader.scala

```diff
@@ -0,0 +1,113 @@
+package at.forsyte.apalache.tla.tracee
+
+import at.forsyte.apalache.infra.passes.options.SourceOption
+import at.forsyte.apalache.infra.passes.options.SourceOption.FileSource
+import at.forsyte.apalache.io.json.impl.{UJsonRep, UJsonScalaFactory, UJsonToTlaViaBuilder}
+import at.forsyte.apalache.infra.passes.options.SourceOption._
+import at.forsyte.apalache.io.json.JsonDeserializationError
+import at.forsyte.apalache.io.lir.TypeTagReader
+import at.forsyte.apalache.tla.imp.src.SourceStore
+import at.forsyte.apalache.tla.lir.oper.{TlaBoolOper, TlaOper}
+import at.forsyte.apalache.tla.lir.{NameEx, OperEx, TlaEx, TlaOperDecl}
+import at.forsyte.apalache.tla.tracee.TraceReader.{ApalacheJson, ITFJson, TraceJson}
+
+import java.io.File
+import scala.annotation.unused
+import scala.util.{Failure, Success, Try}
+
+/**
+ * Reads an ITF-format or an ApalacheJson-format trace into a TLAIR trace, with an intermediate UJson representation.
+ *
+ * The intermediate representation wraps the underlying JSON object with a [[TraceReader.TraceJson]] wrapper, so the
+ * methods know whether to expect the shape of ITF or of ApalacheJson.
+ *
+ * The wrapper can either immediately be passed to [[convert]], or read via e.g. [[getTraceLength]], if we don't need to
+ * convert the full contents of the JSON.
+ *
+ * @author
+ *   Jure Kukovec
+ */
+class UJsonTraceReader(sourceStoreOpt: Option[SourceStore], tagReader: TypeTagReader) extends TraceReader[UJsonRep] {
+  private val builder = new UJsonToTlaViaBuilder(sourceStoreOpt)(tagReader)
+
+  type TraceUJson = TraceJson[UJsonRep]
+
+  // Rethrow as JsonDeserializationError if unable to read
+  private def tryRead(file: File): UJsonRep = Try(ujson.read(file)) match {
+    case Success(ujsonVal) => UJsonRep(ujsonVal)
+    case Failure(exception) =>
+      throw new JsonDeserializationError(s"Unable to read $file as JSON.", exception)
+  }
+
+  override def read(source: SourceOption): TraceUJson = source match {
+    case FileSource(f, Format.Json) => ApalacheJson(tryRead(f))
+    case FileSource(f, Format.Itf)  => ITFJson(tryRead(f))
+    case src => throw new IllegalArgumentException(s"Tried to load state from an invalid source: $src.")
```

Comment:
I think something like this should suffice to support string inputs, if you just change `tryRead` to take a `ujson.Readable`.

```suggestion
    case FileSource(f, Format.Json)     => ApalacheJson(tryRead(f))
    case FileSource(f, Format.Itf)      => ITFJson(tryRead(f))
    case StringSource(s, _, Format.Json => ApalacheJson(tryRead(s))
    case StringSource(s, _, Format.Itf  => ITFJson(tryRead(f))
    case src => throw new IllegalArgumentException(s"Tried to load state from an invalid source: $src.")
```

Reply:
Ah, I didn't realize `StringSource` also carried format info. If that's the case then you're right, we can support it.